### PR TITLE
Alex E - Processors Scaffold + Verify Auth Processor Helper

### PR DESCRIPTION
**Pull request by AlexEntrepreneur**
## Description

Created placeholder processor functions for each endpoint, added promises that return mock data for each endpoint, and implemented an auth verification helper.

Next, I will implement general JSON validation and normalisation helpers + implement the login, register, and `/me` processors.

## Type of change
   
- Scaffolding + Feature Implementation
    
## Change status
- Incomplete
    
## How Has This Been Tested?
       
- Unit Tested + Tested Using VSCode Debugger
    
### Checklist:
    
 - [x] My code follows the style guidelines of this project
 - [x] I have performed a self-review of my own code
 - [ ] My code has been reviewed by at least one peer
 - [x] I have commented my code, particularly in hard-to-understand areas
 - [ ] I have made corresponding changes to the documentation
 - [x] My changes generate no new warnings
 - [x] I have added tests that prove my fix is effective or that my feature works
 - [x] New and existing unit tests pass locally with my changes
 - [x] There are no merge conflicts

### CODE REVIEW

File: server/src/data-processors/helpers/verify-auth.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
+const requiredDependency = require('../../global-helpers/required-dependency');
+const throwError = require('../../global-helpers/throw-error');
+
```
Comment (AnthonyJCampbell) [inline server/src/data-processors/helpers/verify-auth.js:3]: Good call adding the global helpers!

File: server/src/data-processors/helpers/verify-auth.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+function makeVerifyAuth({ jwt = requiredDependency('jwt') } = {}) {
+  const jwtParamHasVerifyMethod = 'verify' in jwt;
+  if (!jwtParamHasVerifyMethod) {
+    requiredDependency('jwt');
+  }
```
Comment (AnthonyJCampbell) [inline server/src/data-processors/helpers/verify-auth.js:8]: What is the added benefit of checking for `verify()`? Doesn't JWT automatically come with `verify()`? Or is it a custom method you've made?

File: server/src/data-processors/helpers/verify-auth.js [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  if (!jwtParamHasVerifyMethod) {
+    requiredDependency('jwt');
+  } 
+  else if (!process.env.AUTHSECRET) {
+    throwError('An "AUTHSECRET" environment variable is required', 'application');
+  } else {
+    return function (token) {
+      let message, tokenBody;
+  
+      if (token && typeof token === 'string') {
```
Comment (AnthonyJCampbell) [inline server/src/data-processors/helpers/verify-auth.js:15]: Nice type-checking! 🔥

File: server/src/data-processors/helpers/verify-auth.js [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        jwt.verify(token, process.env.AUTHSECRET, (err, decodedTokenBody) => {
+          if (err) {
+            throwError('You need authorisation to access this endpoint.', 'authorisation');
+          } else {
+            message = "Successfully authenticated.";
+            tokenBody = decodedTokenBody;
```
Comment (AnthonyJCampbell) [inline server/src/data-processors/helpers/verify-auth.js:21]: Having a spot of trouble figuring out what form decodedTokenBody will have. Will it just be generic string?

File: server/src/data-processors/helpers/verify-auth.js [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+          }
+        });
+  
+        return Object.freeze({
+          message,
+          body: tokenBody
+        });
```
Comment (AnthonyJCampbell) [inline server/src/data-processors/helpers/verify-auth.js:28]: Coming around to the `Object.freeze()`! The more I think about it, the more useful it seems! Awesome work coming up with it! 🥇

File: server/src/data-processors/index.js [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+// TODO: Create processor functions for each endpoint
+
+
+//====== Processors ======//
+// To avoid the app.js from becoming a massive file, dependency injection
+// happens here and then is exported.
+// May refactor the other parts of the app to use a local index file for 
+// dependency injection also depending on the benefits.
+const getMeProcessor = makeInterface({
+  interfaceFunction: getMeProcessorFunction,
```
Comment (AnthonyJCampbell) [inline server/src/data-processors/index.js:22]: Not quite sure what `getMeProcessorFunction` refers to here. Could you run me through it someday soon?

**Review by AnthonyJCampbell — APPROVED** [review]
Awesome work, no changes required as far as I can tell.

File: server/src/data-processors/index.js (hunk 6 above)
Comment (AlexEntrepreneur) [reply server/src/data-processors/index.js:22]: So I named the processor functions after each endpoint `<method><path>ProcessorFunction` (e.g. putResetPasswordProcessorFunction)

File: server/src/data-processors/helpers/verify-auth.js (hunk 2 above)
Comment (AlexEntrepreneur) [reply server/src/data-processors/helpers/verify-auth.js:8]: In future, the `jwt` library used may not necessarily be the `jsonwebtoken` npm package. This gives anyone the ability to use whatever `jwt` dependency they want, as long as it works in the same way with a `verify` method.

File: server/src/data-processors/helpers/verify-auth.js (hunk 4 above)
Comment (AlexEntrepreneur) [reply server/src/data-processors/helpers/verify-auth.js:21]: `decodedTokenBody` will be the payload object from the decoded token. Maybe I should rename it to `decodedTokenPayload`?

**Pull request merged**